packages/grafana-ui/src/components/Typeahead/PartialHighlighter.tsx

```diff
@@ -0,0 +1,55 @@
+import React, { createElement } from 'react';
+import { HighlightPart } from '../../types';
+
+interface Props {
+  text: string;
+  highlightParts: HighlightPart[];
+  highlightClassName: string;
+}
+
+/**
+ * Flattens parts into a list of indices pointing to the index where a part
+ * (highlighted or not highlighted) starts. Adds extra indices if needed
+ * at the beginning or the end to ensure the entire text is covered.
+ */
+function getStartIndices(parts: HighlightPart[], length: number): number[] {
+  const indices: number[] = [];
+  parts.forEach((part) => {
+    indices.push(part.start, part.end + 1);
+  });
+  if (indices[0] !== 0) {
+    indices.unshift(0);
+  }
+  if (indices[indices.length - 1] !== length) {
+    indices.push(length);
+  }
+  return indices;
+}
+
+export const PartialHighlighter: React.FC<Props> = (props: Props) => {
+  let { highlightParts, text, highlightClassName } = props;
+
+  if (!highlightParts) {
+    return null;
+  }
+
+  let children = [];
```

Comment:
What do you think about extracting the child generation logic into a separate function and then memoizing that?

Reply:
I think in this case we won't get much out of memoization. I'd expect the children to be rendered with different props most of the time. The inner parts (`span`/`mark`) are relatively simple and fully generated from PartialHighlighter props so I think it'd wouldn't be more readable if they were extracted into a separate component yet (but definitely it'd be the way to go if more props/logic was added to the inner spans/marks)

Reply:
Cool, makes sense